src/core/bitset.c

```diff
@@ -0,0 +1,649 @@
+    const igraph_integer_t final_block_size = bitset->size % IGRAPH_INTEGER_SIZE ? bitset->size % IGRAPH_INTEGER_SIZE : IGRAPH_INTEGER_SIZE;
+    const igraph_integer_t slots = IGRAPH_BIT_NSLOTS(bitset->size);
+    const igraph_integer_t one = 1, zero = 0; /* to avoid the need to cast 1 and 0 to igraph_integer_t below */
+    const igraph_integer_t mask = final_block_size == IGRAPH_INTEGER_SIZE ? ~zero : ((one << final_block_size) - one);
+
+    for (igraph_integer_t i = 0; i + 1 < slots; ++i) {
+        if (VECTOR(*bitset)[i] != zero) {
+            return IGRAPH_INTEGER_SIZE * i + IGRAPH_CTZ(VECTOR(*bitset)[i]);
+        }
+    }
+    if (bitset->size && (mask & VECTOR(*bitset)[slots - 1]) != zero) {
+        return IGRAPH_INTEGER_SIZE * (slots - 1) + IGRAPH_CTZ(mask & VECTOR(*bitset)[slots - 1]);
+    }
+
+    return bitset->size;
+}
+
+/**
+ * \ingroup bitset
+ * \function igraph_bitset_countr_one
+ * \brief The number of trailing ones in the bitset.
+ *
+ * \experimental
+ *
+ * Returns the number of trailing ones (starting at the least significant bit)
+ * in the bitset before the first zero is encountered.
+ * If the bitset is all ones, then its size is returned.
+ *
+ * \param bitset The bitset object
+ * \return The number of trailing ones in the bitset.
+ *
+ * Time complexity: O(n/w).
+ */
+
+igraph_integer_t igraph_bitset_countr_one(const igraph_bitset_t *bitset) {
+    const igraph_integer_t final_block_size = bitset->size % IGRAPH_INTEGER_SIZE ? bitset->size % IGRAPH_INTEGER_SIZE : IGRAPH_INTEGER_SIZE;
+    const igraph_integer_t slots = IGRAPH_BIT_NSLOTS(bitset->size);
+    const igraph_integer_t one = 1, zero = 0; /* to avoid the need to cast 1 and 0 to igraph_integer_t below */
+    const igraph_integer_t mask = final_block_size == IGRAPH_INTEGER_SIZE ? zero : ~((one << final_block_size) - one);
+
+    for (igraph_integer_t i = 0; i + 1 < slots; ++i) {
+        if (VECTOR(*bitset)[i] != ~zero) {
+            return IGRAPH_INTEGER_SIZE * i + IGRAPH_CTO(VECTOR(*bitset)[i]);
+        }
+    }
+    if (bitset->size && (mask | VECTOR(*bitset)[slots - 1]) != ~zero) {
+        return IGRAPH_INTEGER_SIZE * (slots - 1) + IGRAPH_CTO(mask | VECTOR(*bitset)[slots - 1]);
+    }
+
+    return bitset->size;
+}
+
+/**
+ * \ingroup bitset
+ * \function igraph_bitset_or
+ * \brief Bitwise OR of two bitsets.
+ *
+ * \experimental
+ *
+ * Applies a bitwise or to the contents of two bitsets and stores it in an
+ * already initialized bitset. The destination bitset may be equal to one
+ * (or even both) of the sources.
+ *
+ * \param dest The bitset object where the result is stored
+ * \param src1 A bitset
+ * \param src2 A bitset
+ *
+ * Time complexity: O(n/w).
+ */
+
+void igraph_bitset_or(igraph_bitset_t *dest,
+                      const igraph_bitset_t *src1, const igraph_bitset_t *src2) {
+    for (igraph_integer_t i = 0; i < IGRAPH_BIT_NSLOTS(dest->size); ++i) {
+        VECTOR(*dest)[i] = VECTOR(*src1)[i] | VECTOR(*src2)[i];
+    }
+}
```

Comment:
@ntamas Notice that these bitset functions are different from vector functions in that:

 - `vector_prod(a, b)` does `a *= b`
 - `bitset_or(a, b, c)` does `a = b | c`

An possible advantage is that when one does need the result to go in a new bitset, no copying is necessary. When not, `bitset_or(a, a, b)` is valid. It's unclear how this sort of aliasing interferes with the compiler's ability to optimize, if at all.

The choice between the two kinds of interfaces has an implication on error checking, which is currently missing and should be added. If keeping the current interface, the natural thing to do is to:

 1. check that `src1` and `src2` are of the same size and 
 2. resize `dist` (which is assumed to be initialized) to be the same size

If we do this, that interferes with the ability to use `bitset_or(a, a, b)`. The alternative is to keep these functions very low level and require the user to ensure that sizes are compatible. However, this would be quite different from how other igraph functions work.

----

After thinking this through, I'm leaning towards making the interface be just like `vector_add()`, i.e. `bitset_or(a, b)` should do `a |= b`.

Any comments @ntamas, @Tagl ?

Reply:
There are cases where having three parameters saves you a copy operation. Therefore I believe there should be an option for that. Resizing makes sense to me if we don't leave it to the user to track the sizes. 

Most cases when working with bitsets you specify the size at compile time or runtime but the sizes of all your bitsets are the same. Therefore leaving the size check to the user means that they can simply skip the size checks entirely in most cases.

However, looking at boost::dynamic_bitset, it only provides the in place operators and also says a requirement for calling the functions is the size is equal, so they leave checking to the user.

Reply:
No strong opinions from my side; I don't think we need to keep vectors and bitsets consistent, though, so feel free to do it either way.

Reply:
The decision was to leave this as-is, treating these functions as "low-level". So the interface can stay the same for flexibility, and there won't be any error checks. This should be clearly document, as it will be surprising the users of the library.

@Tagl Can you write a couple of sentences and paste them into the docs of each relevant function please?

Reply:
Updated